### ISSUE TITLE
SRV caching issue

### ISSUE THREAD

**Issue by DL6ER** [pi-hole/FTL#849]
I think I have encountered this or a similar bug (specifically with a BT cloud voice express phone I was trying to fix).

pihole.log
Jul 14 23:18:01 dnsmasq[1105]: query[NAPTR] ipcomms-btb-centrex-wv20lnws20.bt.com from 10.11.1.89 (src/dnsmasq/forward.c:1543)
Jul 14 23:18:01 dnsmasq[1105]: forwarded ipcomms-btb-centrex-wv20lnws20.bt.com to 127.0.0.1 (src/dnsmasq/forward.c:547)
Jul 14 23:18:01 dnsmasq[1105]: query[SRV] _sip._udp.ipcomms-btb-centrex-wv20lnws20.bt.com from 10.11.1.89 (src/dnsmasq/forward.c:1543)
Jul 14 23:18:01 dnsmasq[1105]: config _sip._udp.ipcomms-btb-centrex-wv20lnws20.bt.com is NODATA (src/dnsmasq/rfc1035.c:1969)
Jul 14 23:18:01 dnsmasq[1105]: query[A] ipcomms-btb-centrex-wv20lnws20.bt.com from 10.11.1.89 (src/dnsmasq/forward.c:1543)
Jul 14 23:18:01 dnsmasq[1105]: cached ipcomms-btb-centrex-wv20lnws20.bt.com is NODATA-IPv4 (src/dnsmasq/rfc1035.c:1794)

and for ftl:
[2020-07-14 23:18:01.719 1105] *************************** unknown CACHE reply (2) ***************************

web ui will show something like this:

2020-07-14 23:30:28 | NAPTR | ipcomms-btb-centrex-wv20lnws20.bt.com | queryinghost | OK (forwarded) | N/A | 
2020-07-14 23:29:41 | A | ipcomms-btb-centrex-wv20lnws20.bt.com | queryinghost | OK (cached) | NODATA (0.2ms) | 
2020-07-14 23:29:41 | SRV | _sip._udp.ipcomms-btb-centrex-wv20lnws20.bt.com | queryinghost | Unknown (0) | 

Changing the upstream server in pihole from its local stubby dns-over-tls proxy to cloudflare, google or opendns directly also still results in NODATA being returned.

I have checked out fix/process_SRV_cache_master as suggested in https://github.com/pi-hole/FTL/issues/759 but still have the same behaviour.

an nslookup from a windows machine via pihole returns
*** No Service location (SRV) records available for _sip._udp.ipcomms-btb-centrex-wv20lnws20.bt.com

querying e.g. cloudflare directly returns the correct result
_sip._udp.ipcomms-btb-centrex-wv20lnws20.bt.com SRV service location:
          priority       = 20
          weight         = 20
          port           = 5060
          svr hostname   = ipcomms-btb-centrex-lnws20.bt.com
_sip._udp.ipcomms-btb-centrex-wv20lnws20.bt.com SRV service location:
          priority       = 10
          weight         = 20
          port           = 5060
          svr hostname   = ipcomms-btb-centrex-wv20.bt.com

As these are not local only srv records relating to AD domains (as these 2 issues appear to refer to) it should hopefully be easy to reproduce for the developers.

This problem does not occur for me with dnsmasq 2.80 on openwrt at least and configuring the device with alternate dns causes to succeed in registration (the records retreived when bypassing the pihole are valid)!

_Originally posted by @undesiredpopup in https://github.com/pi-hole/pi-hole/issues/3399#issuecomment-658448927_

**Comment by DL6ER** [comment]
I can confirm this. It is a bug in `dnsmasq`'s SRV caching.

For reference, the first lookup gives
```
$ dig SRV _sip._udp.ipcomms-btb-centrex-wv20lnws20.bt.com

; <<>> DiG 9.11.3-1ubuntu1.12-Ubuntu <<>> SRV _sip._udp.ipcomms-btb-centrex-wv20lnws20.bt.com
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 31005
;; flags: qr rd ra; QUERY: 1, ANSWER: 2, AUTHORITY: 4, ADDITIONAL: 7

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1472
;; QUESTION SECTION:
;_sip._udp.ipcomms-btb-centrex-wv20lnws20.bt.com. IN SRV

;; ANSWER SECTION:
_sip._udp.ipcomms-btb-centrex-wv20lnws20.bt.com. 600 IN	SRV 20 20 5060 ipcomms-btb-centrex-lnws20.bt.com.
_sip._udp.ipcomms-btb-centrex-wv20lnws20.bt.com. 600 IN	SRV 10 20 5060 ipcomms-btb-centrex-wv20.bt.com.

;; AUTHORITY SECTION:
bt.com.			86400	IN	NS	EDDNS0.bt.com.
bt.com.			86400	IN	NS	DYDNS0.bt.com.
bt.com.			86400	IN	NS	EDDNS1.bt.com.
bt.com.			86400	IN	NS	DYDNS1.bt.com.

;; ADDITIONAL SECTION:
ipcomms-btb-centrex-wv20.bt.com. 600 IN	A	217.32.186.139
ipcomms-btb-centrex-lnws20.bt.com. 600 IN A	217.32.186.75
EDDNS0.bt.com.		86400	IN	A	193.113.57.242
EDDNS1.bt.com.		86400	IN	A	193.113.57.243
DYDNS0.bt.com.		86400	IN	A	193.113.32.156
DYDNS1.bt.com.		86400	IN	A	193.113.32.157

;; Query time: 155 msec
;; SERVER: 192.168.2.10#53(192.168.2.10)
;; WHEN: Mon Jul 20 23:24:36 CEST 2020
;; MSG SIZE  rcvd: 412
```
and the SRV record is added to `dnsmasq`'s cache.

Any subsequent (within the TTL lifetime) query results in `NODATA`:
```
$ dig SRV _sip._udp.ipcomms-btb-centrex-wv20lnws20.bt.com

; <<>> DiG 9.11.3-1ubuntu1.12-Ubuntu <<>> SRV _sip._udp.ipcomms-btb-centrex-wv20lnws20.bt.com
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 30531
;; flags: qr rd ra; QUERY: 1, ANSWER: 2, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;_sip._udp.ipcomms-btb-centrex-wv20lnws20.bt.com. IN SRV

;; ANSWER SECTION:
_sip._udp.ipcomms-btb-centrex-wv20lnws20.bt.com. 575 IN	SRV 10 20 5060 ipcomms-btb-centrex-wv20.bt.com.
_sip._udp.ipcomms-btb-centrex-wv20lnws20.bt.com. 575 IN	SRV 20 20 5060 ipcomms-btb-centrex-lnws20.bt.com.

;; Query time: 1 msec
;; SERVER: 192.168.2.10#53(192.168.2.10)
;; WHEN: Mon Jul 20 23:25:01 CEST 2020
;; MSG SIZE  rcvd: 180
```

**Comment by DL6ER** [comment]
@undesiredpopup I cannot reproduce the
```
*************************** unknown CACHE reply (2) ***************************
```
line. The line following this one is of interest as it shows what the flags of this query were.

Which version of Pi-hole FTL were you using at that time?

It is likely that this is a bug in `dnsmasq` which needs to be reported upstream. I just confirmed that this is still present with the most recent v2.82. It makes sense that this was not an issue in dnsmasq v2.80 as SRV caching is a new feature in v2.81 (first released in Pi-hole v5.0).

**Comment by DL6ER** [comment]
Closing due to inactivity. Feel free to reopen if the issue persists.
